### ISSUE TITLE
Cart session data is not validated

### ISSUE THREAD

**Issue by tristanlins** [isotope/core#1334]
- Put some products into the cart.
- Add a coupon code 
- Continue the purchase process until the last step
- Stop here, go back to the cart and remove all products
- Put another product into the cart (the product must excluded from delivery and the coupon is not compatible to the new products)
- Continue the purchase process
- The delivery and coupon are still active
![isotope](https://cloud.githubusercontent.com/assets/343404/5300646/2dc42b2c-7bd2-11e4-91e4-918afd58e019.png)

**Comment by aschempp** [comment]
I can't reproduce this in 2.1, I assume @tristanlins was using 2.2? Because there it would make sense with the new darft order…

**Comment by tristanlins** [comment]
No, it is a 2.1 from ER2.

**Comment by aschempp** [comment]
Unfortunately, I can't reproduce the problem. Let me know if you're still working on that project or who could assist me with. I'll close the issue in the meantime.

**Comment by kuvb** [comment]
+1
I just tried and could easily reproduce it with the demo installation (just updated to 2.1.10 Build 7)

![isotope-checkout-download-shipping](https://cloud.githubusercontent.com/assets/5165551/6082253/4b11090e-ae20-11e4-91b6-7124d4abff9b.jpg)

**Comment by tristanlins** [comment]
My customer tell me last week, he can reproduce this with the official isotope demo! I will have a look today for this and maybe I can provide a fix later.

**Comment by tristanlins** [comment]
Please see https://github.com/isotope/core/pull/1363

**Comment by kuvb** [comment]
The label "invalid" might be a little confusing :)
I used the following demo files of Isotope: isotope_demo_2.1.7.zip. Happened before and after the update.
With tristanlins pull, it gets validated correctly on the review page, however not in the shopping cart.

You can reproduce the behaviour the following way:

- Add download product and shipping product to cart
- You need a shipping method that costs some money of course
- Proceed to review page
- Remove shipping product from cart
=> You will only have one download product in the cart and the price for shipping added as shown above in the screenshots.

**Comment by kuvb** [comment]
@aschempp Were you able to reproduce this?
Is there a more "solid" fix that I can apply? The one of tristanlins only removes the shipping charges on the review page, not in the shopping cart.

**Comment by tristanlins** [comment]
We did a quick modification on the `Cart.php:87` to solve the problem that an unnecessary shipping is displayed in the card (the other half of #1363):

```php
          // Remove shipping if not required
          if (!Isotope::getCart()->requiresShipping()) {
              Isotope::getCart()->setShippingMethod(null);
          }
```

This works, but in my eyes this is not a good solution...

**Comment by aschempp** [comment]
I suggest you do upgrade to version 2.2. Unfortunately, this can't be correctly fixed in 2.1. For this reason, we have removed shipping information from Cart in 2.1 and restructured the checkout process. Let me know if the problem persists in 2.2.
